did_peer/src/numalgos/numalgo2/generate/helpers.rs

```diff
@@ -0,0 +1,256 @@
+use base64::{engine::general_purpose::STANDARD_NO_PAD, Engine};
+use did_doc::schema::{did_doc::DidDocument, verification_method::VerificationMethodKind};
+use did_doc_sov::extra_fields::ExtraFieldsSov;
+
+use crate::{
+    error::DidPeerError,
+    key::get_key_by_verification_method,
+    numalgos::numalgo2::{purpose::ElementPurpose, service_abbreviated::ServiceAbbreviated},
+};
+
+pub fn append_encoded_key_segments(
+    mut did: String,
+    did_document: &DidDocument<ExtraFieldsSov>,
+) -> Result<String, DidPeerError> {
+    for am in did_document.assertion_method() {
+        did = append_encoded_key_segment(did, &did_document, am, ElementPurpose::Assertion)?;
+    }
+
+    for ka in did_document.key_agreement() {
+        did = append_encoded_key_segment(did, &did_document, ka, ElementPurpose::Encryption)?;
+    }
+
+    for vm in did_document.verification_method() {
+        did = append_encoded_key_segment(
+            did,
+            &did_document,
+            &VerificationMethodKind::Resolved(vm.to_owned()),
+            ElementPurpose::Verification,
+        )?;
+    }
+
+    for a in did_document.authentication() {
+        did = append_encoded_key_segment(did, &did_document, a, ElementPurpose::Verification)?;
+    }
+
+    for ci in did_document.capability_invocation() {
+        did = append_encoded_key_segment(did, &did_document, ci, ElementPurpose::CapabilityInvocation)?;
+    }
+
+    for cd in did_document.capability_delegation() {
+        did = append_encoded_key_segment(did, &did_document, cd, ElementPurpose::CapabilityDelegation)?;
+    }
+
+    Ok(did)
+}
+
+pub fn append_encoded_service_segment(
+    mut did: String,
```

Comment:
Since this only appends, why not take a `&mut String`?

Reply:
I strongly prefer immutability wherever possible, and following the principle of least surprise, i.e. in this case taking the old version by value and returning new version of the string instead of mutating the argument "silently" in place.

Reply:
I'd argue that it is less surprising to pass in a `&mut String` and realize it will get mutated rather than passing in an owned `String` just to get it back after it gets mutated. Because whichever way it's done, the very same heap memory location *does* get mutated.

Reply:
I don't see how the fact that the very same heap memory location gets mutated both ways influences the degree of surprise, or how is it argument either way. What I meant by surprise to avoid are the potential side-effects related to passing around mutable references (making it unclear where the actual value gets mutated). Again, I prefer passing ownership and returning new value - it avoids creating mutable references and side-effects, it clearly establishes ownership of the value and the places where it is modified.

Reply:
Alright, your PR, your call.